static-checks/ansible/allowed-modules/test.py

```diff
@@ -0,0 +1,134 @@
+import subprocess
+import tempfile
+import yaml
+import contextlib
+from pathlib import Path
+
+from lib import util, results
+
+
+# Obtained from
+# https://docs.ansible.com/ansible/latest/reference_appendices/playbooks_keywords.html#task
+ansible_reserved_keywords = ['action', 'any_errors_fatal', 'args', 'async', 'become', 'become_exe',
+                             'become_flags', 'become_method', 'become_user', 'changed_when',
+                             'check_mode', 'collections', 'connection', 'debugger', 'delay',
+                             'delegate_facts', 'delegate_to', 'diff', 'environment', 'failed_when',
+                             'ignore_errors', 'ignore_unreachable', 'local_action', 'loop',
+                             'loop_control', 'module_defaults', 'name', 'no_log', 'notify', 'poll',
+                             'port', 'register', 'remote_user', 'retries', 'run_once', 'tags',
+                             'throttle', 'timeout', 'until', 'vars', 'when', 'with_items']
+
+
+@contextlib.contextmanager
+def select_all_rules(datastream):
+    """
+    Given an XML file path as 'datastream' enable all the rules in the datastream.
+
+    Use as a context manager, returning a filesystem path to the
+    modified datastream.
+    """
+    exprs = set()
+    exprs.add(re.compile(r'<.*Rule.*id="'))
+    exprs.add(re.compile(r'<.*select.*idref="'))
+
+    with open(datastream) as ds:
+        with tempfile.NamedTemporaryFile(mode='w', suffix='.xml') as new_ds:
+            for line in ds:
+                if any(x.search(line) for x in exprs):
+                    line = line.replace('selected="false"', 'selected="true"')
+                new_ds.write(line)
+            new_ds.flush()
+            util.log(f"Saving modified {datastream} with all rules being selected as"
+                     "{new_ds.name}")
+            yield Path(new_ds.name)
+
+
+def list_append_ansible_module_from_task(modules_list, task_keywords, allowed_modules):
+    updated_keywords = set()
+    for keyword in task_keywords:
+        keyword = keyword.replace("ansible.builtin.", "")
+        updated_keywords.add(keyword)
+    task_module = list(updated_keywords.intersection(allowed_modules))
+    if len(task_module) == 1:
+        module = str(task_module[0])
+        if modules_list.count(module) == 0:
+            # Module module is not yet in the list, append it.
+            modules_list.append(module)
+    else:
+        module = ''.join([x for x in list(updated_keywords) if x not in
+                          ansible_reserved_keywords])
+        results.report('fail', module, 'module is not allowed')
+        return 1
+    return 0
+
+
+def traverse_blocks(modules_list, task, allowed_modules):
+    block_set = set(['block'])
+    task_keywords = set(task.keys())
+
+    if len(task_keywords.intersection(block_set)) > 0:
+        for block_task in task['block']:
+            return traverse_blocks(modules_list, block_task, allowed_modules)
```

Comment:
Same as for traversing tasks in the parent function, `return` here (and in the `else` block) doesn't make sense with the `for`. Maybe you meant to `yield` blocks to parent or not `return` at all?

Reply:
Yeah, good catch!